### PR TITLE
Add top-k post-processing to RT-DETR decoder head

### PR DESCRIPTION
**Pull request by artest08**
Replaces the per-query argmax in `RTDETRValidator.postprocess` and `RTDETRPredictor.postprocess` with a top-k over flattened class scores inside `RTDETRDecoder.postprocess`. A `disable_topk` class flag preserves the previous argmax behavior. Decoder inference now returns `(bs, num_queries, 6)` as `[cx, cy, w, h, score, class]`, so val/predict only convert and scale boxes.

## Summary

Move RT-DETR top-k post-processing from the validator/predictor into the decoder head, matching the convention used by DETR-family detectors (DETR, Deformable-DETR, DINO, the original RT-DETR reference impl). Selection now runs over the flattened `(num_queries × num_classes)` score grid instead of reducing each query to a single argmax class, which lets the model emit multiple predictions per query for different classes.

## Changes

- **`ultralytics/nn/modules/head.py`**
  - Add `RTDETRDecoder.postprocess(boxes, scores)` that performs `scores.flatten(1).topk(num_queries)` and returns `(bs, num_queries, 6)` as `[cx, cy, w, h, score, class]`.
  - Inference path of `RTDETRDecoder.forward` now calls `postprocess` and returns the `(bs, num_queries, 6)` tensor (previously `(bs, num_queries, 4 + nc)`).
  - Add `disable_topk` class attribute (default `False`) to fall back to per-query argmax for users who want the previous behavior.
- **`ultralytics/models/rtdetr/val.py`**
  - `RTDETRValidator.postprocess` simplified to xywh→xyxy + image-size scaling; selection is already done in the head.
- **`ultralytics/models/rtdetr/predict.py`**
  - `RTDETRPredictor.postprocess` updated to consume the new 6-channel output (drops the local argmax + score-sort + max-det slice, since top-k is done upstream).

## Why

Top-k over the joint `(query, class)` distribution is the de-facto standard in transformer-based detectors. The previous argmax-per-query path discarded valid secondary class predictions per query, which is why this change recovers a measurable mAP gain without any training change.

## Results

COCO val2017, `imgsz=640`, `half=True`, weights from `ultralytics/assets`:

| Model     | Before (argmax) | After (top-k) | Δ mAP50-95 |
|-----------|-----------------|---------------|------------|
| rtdetr-l  | 52.3            | 52.7          | +0.4       |
| rtdetr-x  | 54.0            | 54.4          | +0.4       |

## Backwards compatibility

- The decoder inference output tensor shape changed from `(bs, num_queries, 4 + nc)` to `(bs, num_queries, 6)`. Downstream code inside this repo (`RTDETRValidator`, `RTDETRPredictor`) is updated accordingly. External code that consumes raw decoder output will need to adapt.
- Setting `RTDETRDecoder.disable_topk = True` restores the previous argmax-per-query behavior.

## Test plan

- [x] `yolo val model=rtdetr-l.pt data=coco.yaml half=True` → mAP50-95 = **52.7** ✅
- [x] `yolo val model=rtdetr-x.pt data=coco.yaml half=True` → mAP50-95 = **54.4** ✅
- [x] `yolo predict model=rtdetr-l source=bus.jpg` — PyTorch `.pt` and TensorRT `.engine` produce identical detections: **4 persons, 1 bus, 1 traffic light, 1 fire hydrant, 2 ties**
- [x] TensorRT validation: `yolo val model=rtdetr-l.engine data=coco.yaml half=True` → mAP50-95 = **52.7** (matches PyTorch)

> ⚠️ **Breaking change for cached TensorRT engines.** The decoder inference output shape changed from `(bs, num_queries, 4 + nc)` to `(bs, num_queries, 6)`, so previously exported `.engine` (and any other format with a baked-in output graph) files are no longer compatible and **must be re-exported** from the `.pt` weights.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
✨ This PR streamlines RT-DETR inference and validation by moving top-k detection selection into the model head and standardizing outputs to a simpler `[box, score, class]` format.

### 📊 Key Changes
- 🧠 Added a new `postprocess()` method in the RT-DETR head to select the top detections directly inside the decoder head.
- 📦 Changed inference output format from raw class-score tensors to a compact shape of `(batch, num_queries, 6)`:
  - bounding box
  - confidence score
  - class label
- 🚀 Updated `rtdetr/predict.py` to consume these pre-filtered predictions directly, instead of recomputing the best class per box afterward.
- 🧹 Simplified `rtdetr/val.py` postprocessing:
  - now treats predictions as already top-k selected
  - converts boxes from normalized `xywh` to pixel `xyxy`
  - applies confidence filtering more directly
- 📝 Improved docstrings and comments to reflect the new prediction format and processing flow.

### 🎯 Purpose & Impact
- ⚡ Reduces duplicate postprocessing work during prediction and validation, making the pipeline cleaner and potentially faster.
- ✅ Makes RT-DETR outputs more consistent and easier to use across inference, export, and validation.
- 🔍 Lowers the chance of mismatches between model head output and downstream postprocessing logic.
- 🛠️ Simplifies maintenance for developers by centralizing top-k selection in one place.
- 👥 For users, this should mean more predictable RT-DETR behavior with no workflow changes needed, but with a more robust internal implementation.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @artest08, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review, and an engineer will assist you soon. Please review the checklist below to help ensure smooth integration ✨

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/models/rtdetr/predict.py [hunk 1]
```diff
@@ -51,21 +52,18 @@ def postprocess(self, preds, img, orig_imgs):
...
+        for bbox, score, label, orig_img, img_path in zip(bboxes, scores, labels, orig_imgs, self.batch[0]):
             bbox = ops.xywh2xyxy(bbox)
-            max_score, cls = score.max(-1, keepdim=True)  # (300, 1)
-            idx = max_score.squeeze(-1) > self.args.conf  # (300, )
+            idx = score.squeeze(-1) > self.args.conf
             if self.args.classes is not None:
-                idx = (cls == torch.tensor(self.args.classes, device=cls.device)).any(1) & idx
-            pred = torch.cat([bbox, max_score, cls], dim=-1)[idx]  # filter
-            pred = pred[pred[:, 4].argsort(descending=True)][: self.args.max_det]
+                idx = (label.squeeze(-1) == torch.tensor(self.args.classes, device=label.device)).any(1) & idx
```
Comment (UltralyticsAssistant) [inline ultralytics/models/rtdetr/predict.py]: ⚠️ **HIGH**: `label.squeeze(-1)` breaks the class filter broadcast. `label` comes in as `(num_queries, 1)`, which compares correctly against a `(k,)` class tensor, but after squeezing it becomes `(num_queries,)`; multi-class filters will raise a size-mismatch error, and single-class filters will make `.any(1)` fail because dimension 1 no longer exists.

**Suggested change:**
```suggestion
                idx = (label == torch.tensor(self.args.classes, device=label.device)).any(1) & idx
```

File: ultralytics/models/rtdetr/val.py [hunk 2]
```diff
@@ -158,35 +158,31 @@ def scale_preds(self, predn: dict[str, torch.Tensor], pbatch: dict[str, Any]) ->
...
-            pred = torch.cat([bbox, score[..., None], cls[..., None]], dim=-1)  # filter
-            # Sort by confidence to correctly get internal metrics
-            pred = pred[score.argsort(descending=True)]
-            outputs[i] = pred[score > self.args.conf]
-
-        return [{"bboxes": x[:, :4], "conf": x[:, 4], "cls": x[:, 5]} for x in outputs]
+        bboxes, scores, labels = preds[0].split((4, 1, 1), dim=-1)
+        bboxes = ops.xywh2xyxy(bboxes) * self.args.imgsz
+
+        return [
```
Comment (UltralyticsAssistant) [inline ultralytics/models/rtdetr/val.py:184]: ⚠️ **HIGH**: The previous validator applied `self.args.conf` before returning predictions for metrics/JSON export. Returning every top-k candidate here changes validation semantics and will include many low-confidence boxes as false positives. This should still mask by confidence when building each per-image output.

File: ultralytics/models/rtdetr/predict.py [hunk 3]
```diff
@@ -51,21 +52,18 @@ def postprocess(self, preds, img, orig_imgs):
...
+            pred = torch.cat([bbox, score, label], dim=-1)[idx]
```
Comment (UltralyticsAssistant) [inline ultralytics/models/rtdetr/predict.py]: 💡 **MEDIUM**: This drops `max_det` enforcement in prediction. The decoder still returns `num_queries` candidates, so when users set `max_det < num_queries` they'll now receive more detections than requested. Since the head-side top-k output is already score-sorted, you can preserve the old behavior by slicing here after filtering.

**Suggested change:**
```suggestion
            pred = torch.cat([bbox, score, label], dim=-1)[idx][: self.args.max_det]
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The decoder-side top-k postprocessing looks directionally good, but there are three downstream regressions to fix before merging: predictor class filtering currently errors, `max_det` is no longer honored during prediction, and validator confidence filtering was removed.

💬 Posted 3 inline comments

File: ultralytics/models/rtdetr/predict.py (hunk 3 above)
Comment (artest08) [reply ultralytics/models/rtdetr/predict.py]: This is integrated as suggested!

File: ultralytics/models/rtdetr/predict.py (hunk 1 above)
Comment (artest08) [reply ultralytics/models/rtdetr/predict.py]: This is also fixed as suggested, now this works
yolo predict model=rtdetr-l.pt source=~/git/ultralytics2/bus.jpg half=True "classes=[0,5]"

File: ultralytics/models/rtdetr/val.py (hunk 2 above)
Comment (artest08) [reply ultralytics/models/rtdetr/val.py:184]: I have also integrated self.args.conf here, but is it very crucial, why are we utilizing different thresholds for evaluation pipelines, I think for coco eval pipeline, generally integrating all predictions yields the best result right?

**Review by Laughing-q — APPROVED** [review]

**Comment by Laughing-q** [comment]
@artest08 looks good! I removed the `disable_topk` option since it's not used

**Comment by UltralyticsAssistant** [comment]
🎉 Merged! Huge thanks to @artest08 for this thoughtful RT-DETR cleanup, with contributions from @Laughing-q.

As Leonardo da Vinci said, *“Simplicity is the ultimate sophistication.”* This PR reflects that well by moving top-k selection into the head and standardizing outputs to a cleaner `[box, score, class]` format, making RT-DETR inference and validation simpler, more consistent, and easier to maintain.

Really appreciate the care put into reducing duplicated postprocessing and improving the overall prediction flow. Great work by both of you 🚀
